cedar-policy-validator/src/human_schema/err.rs

```diff
@@ -354,80 +363,324 @@ impl Diagnostic for ToJsonSchemaErrors {
+    fn labels(&self) -> Option<Box<dyn Iterator<Item = LabeledSpan> + '_>> {
+        let l1 = LabeledSpan::underline(self.loc1.span);
+        let l2 = LabeledSpan::underline(self.loc2.span);
+        Some(Box::new(vec![l1, l2].into_iter()))
+    }
+}
+
+/// Error raised when there are duplicate `principal` or `resource` declarations
+#[derive(Debug, Clone, Error, PartialEq, Eq)]
+#[error("duplicate `{kind}` declaration. Action may have at most one {kind} declaration")]
+pub struct DuplicatePrincipalOrResource {
+    kind: PR,
+    loc1: Loc,
+    loc2: Loc,
+}
+
+impl DuplicatePrincipalOrResource {
+    #[cfg(test)]
+    pub(crate) fn kind(&self) -> PR {
+        self.kind
+    }
+}
+
+impl Diagnostic for DuplicatePrincipalOrResource {
+    fn labels(&self) -> Option<Box<dyn Iterator<Item = LabeledSpan> + '_>> {
+        let l1 = LabeledSpan::underline(self.loc1.span);
+        let l2 = LabeledSpan::underline(self.loc2.span);
+        Some(Box::new(vec![l1, l2].into_iter()))
+    }
+}
+
+/// Error raised when there are duplicate namespace IDs
+#[derive(Debug, Clone, Error, PartialEq, Eq)]
+#[error("Duplicate namespace IDs: `{namespace_id}`")]
+pub struct DuplicateNameSpaces {
+    namespace_id: SmolStr,
+    loc1: Option<Loc>,
+    loc2: Option<Loc>,
+}
+
+impl DuplicateNameSpaces {
+    pub fn namespace_id(&'_ self) -> impl AsRef<str> + '_ {
+        self.namespace_id.as_ref()
+    }
+}
+
+impl Diagnostic for DuplicateNameSpaces {
+    fn labels(&self) -> Option<Box<dyn Iterator<Item = LabeledSpan> + '_>> {
+        let mut v = vec![];
+
+        if let Some(l1) = self.loc1.as_ref() {
+            v.push(LabeledSpan::underline(l1.span));
+        }
+
+        if let Some(l2) = self.loc2.as_ref() {
+            v.push(LabeledSpan::underline(l2.span));
+        }
+
+        Some(Box::new(v.into_iter()))
+    }
+}
+/// Error raised when a type is used without being defined
+#[derive(Debug, Clone, Error, PartialEq, Eq)]
+#[error("Unknown type name: `{}`", .0.node)]
+pub struct UnknownTypeName(Node<SmolStr>);
+
+impl UnknownTypeName {
+    /// Get the name that was invalid
+    pub fn name(&'_ self) -> impl AsRef<str> + '_ {
+        self.0.node.as_ref()
+    }
+}
+
+impl Diagnostic for UnknownTypeName {
+    fn labels(&self) -> Option<Box<dyn Iterator<Item = LabeledSpan> + '_>> {
+        Some(Box::new(
+            vec![LabeledSpan::underline(self.0.loc.span)].into_iter(),
+        ))
+    }
+}
+
+/// Use of a reserved namespace
+#[derive(Debug, Clone, Error, PartialEq, Eq)]
+#[error("Use reserved namespace `__cedar`")]
+pub struct UseReservedNamespace(Loc);
+
+impl Diagnostic for UseReservedNamespace {
+    fn labels(&self) -> Option<Box<dyn Iterator<Item = LabeledSpan> + '_>> {
+        Some(Box::new(
+            vec![LabeledSpan::underline(self.0.span)].into_iter(),
+        ))
+    }
 }
 
 impl ToJsonSchemaError {
+    /// Creates a duplicate keys error
     pub fn duplicate_keys(key: SmolStr, loc1: Loc, loc2: Loc) -> Self {
```

Comment:
Since this type is publicly exposed, end-users can see and use these constructors, and we have to commit to stability for their arguments.  I think it would be better if we could avoid (at least fully `pub`) constructors, so that in the future we could add or change arguments as necessary?

Reply:
agreed. I think our error constructors are generally not public